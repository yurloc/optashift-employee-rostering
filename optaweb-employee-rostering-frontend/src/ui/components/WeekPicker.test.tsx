/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
// @ts-ignore
import DatePicker from '@wojtekmaj/react-daterange-picker';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import moment from 'moment-timezone';
import 'moment/locale/en-ca';

import MockDate from 'mockdate';
import WeekPicker from './WeekPicker';

describe('WeekPicker component', () => {
  beforeAll(() => {
    moment.locale('en');
  });

  it('should render correctly when closed', () => {
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={jest.fn()} />);
    expect(toJson(weekPicker)).toMatchSnapshot();
  });

  it('should render correctly when opened', () => {
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={jest.fn()} />);
    weekPicker.setState({ isOpen: true });
    expect(toJson(weekPicker)).toMatchSnapshot();
  });

  it('should pass previous week to onChange when previous week is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.find('Button[aria-label="Previous Week"]').simulate('click');
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-06-23').toDate(), moment('2019-06-23').endOf('week').toDate());
  });

  it('should pass next week to onChange when next week is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.find('Button[aria-label="Next Week"]').simulate('click');
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-07-07').toDate(), moment('2019-07-07').endOf('week').toDate());
  });

  it('should go to current week when the reset button is clicked', () => {
    const solvingStartTime = moment('2019-07-23').toDate();
    MockDate.set(solvingStartTime);
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    ((weekPicker.find(DatePicker).props() as any).onChange as Function)({} as React.FormEvent);
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-07-21').toDate(), moment('2019-07-21').endOf('week').toDate());
  });

  it('should go to the week containing a day when the day is clicked', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    ((weekPicker.find(DatePicker).props() as any).onClickDay as Function)(moment('2019-07-30').toDate());
    expect(onChange).toBeCalled();
    expect(onChange)
      .toBeCalledWith(moment('2019-07-28').toDate(), moment('2019-07-28').endOf('week').toDate());
  });

  it('should open when opened', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    ((weekPicker.find(DatePicker).props() as any).onCalendarOpen as Function)();
    expect(weekPicker.instance().state).toEqual({
      isOpen: true,
    });
  });

  it('should closed when closed', () => {
    const onChange = jest.fn();
    const weekPicker = shallow(<WeekPicker value={moment('2019-07-03').toDate()} onChange={onChange} />);
    weekPicker.setState({ isOpen: true });
    ((weekPicker.find(DatePicker).props() as any).onCalendarClose as Function)();
    expect(weekPicker.instance().state).toEqual({
      isOpen: false,
    });
  });
});
