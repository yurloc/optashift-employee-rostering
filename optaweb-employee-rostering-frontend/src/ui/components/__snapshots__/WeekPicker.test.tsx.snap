// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekPicker component should render correctly when closed 1`] = `
<div
  className="week-picker-container"
>
  <Button
    aria-label="Previous Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
        transform=""
      />
    </svg>
  </Button>
  <DateRangePicker
    calendarIcon={
      <svg
        className="react-daterange-picker__calendar-button__icon react-daterange-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    className="week-picker"
    clearIcon={
      <HistoryIcon
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
        title={null}
      />
    }
    isOpen={false}
    locale="en-US"
    name="daterange"
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickDay={[Function]}
    required={true}
    value={
      Array [
        2019-06-30T00:00:00.000Z,
        2019-07-06T23:59:59.999Z,
      ]
    }
  />
  <Button
    aria-label="Next Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
        transform=""
      />
    </svg>
  </Button>
</div>
`;

exports[`WeekPicker component should render correctly when opened 1`] = `
<div
  className="week-picker-container"
>
  <Button
    aria-label="Previous Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
        transform=""
      />
    </svg>
  </Button>
  <DateRangePicker
    calendarIcon={
      <svg
        className="react-daterange-picker__calendar-button__icon react-daterange-picker__button__icon"
        height={19}
        stroke="black"
        strokeWidth={2}
        viewBox="0 0 19 19"
        width={19}
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="15"
          width="15"
          x="2"
          y="2"
        />
        <line
          x1="6"
          x2="6"
          y1="0"
          y2="4"
        />
        <line
          x1="13"
          x2="13"
          y1="0"
          y2="4"
        />
      </svg>
    }
    className="week-picker"
    clearIcon={
      <HistoryIcon
        color="currentColor"
        noVerticalAlign={false}
        size="sm"
        title={null}
      />
    }
    isOpen={true}
    locale="en-US"
    name="daterange"
    onCalendarClose={[Function]}
    onCalendarOpen={[Function]}
    onChange={[Function]}
    onClickDay={[Function]}
    required={true}
    value={
      Array [
        2019-06-30T00:00:00.000Z,
        2019-07-06T23:59:59.999Z,
      ]
    }
  />
  <Button
    aria-label="Next Week"
    onClick={[Function]}
    variant="plain"
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      role="img"
      style={
        Object {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 256 512"
      width="1em"
    >
      <path
        d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
        transform=""
      />
    </svg>
  </Button>
</div>
`;
